### PR TITLE
improvements for StringUtils::StringUTF8

### PR DESCRIPTION
**Pull request by summerinsects**
1. add 'const'
2. add range support for 'getAsCharSequence'

### CODE REVIEW

File: cocos/base/ccUTF8.h [hunk 1]
```diff
@@ -207,13 +207,14 @@ class CC_DLL StringUTF8
     std::size_t length() const;
     void replace(const std::string& newStr);
 
-    std::string getAsCharSequence() const;
+    std::string getAsCharSequence(std::size_t pos = 0, std::size_t len = (std::size_t)-1) const;
```
Comment (perminovVS) [inline cocos/base/ccUTF8.h]: Please, use C++ coding style for cocos2d-x
https://github.com/cocos2d/cocos2d-x/blob/v3/docs/CODING_STYLE.md#default-arguments
Not use default arguments, add new function.

**Review by perminovVS — CHANGES_REQUESTED** [review]

File: cocos/base/ccUTF8.h (hunk 1 above)
Comment (summerinsects) [reply cocos/base/ccUTF8.h]: OK

File: cocos/base/ccUTF8.h [hunk 2]
```diff
@@ -196,7 +196,7 @@ class CC_DLL StringUTF8
     struct CharUTF8
     {
         std::string _char;
-        bool isAnsi() { return _char.size() == 1; }
+        bool isAnsi() const { return _char.size() == 1; }
```
Comment (minggo) [inline cocos/base/ccUTF8.h]: utf8 chars' size may be 1 too, right?

Comment (summerinsects) [reply cocos/base/ccUTF8.h]: chars' size is 1 for ansi char, maybe great than 1 other cases

Comment (minggo) [reply cocos/base/ccUTF8.h]: Yep, what i mean is that

* size > 1, not ansi
* size = 1, may be ansi or not

so can not use it to determine if it is ansi or not

Comment (summerinsects) [reply cocos/base/ccUTF8.h]: so the function name is not accurate. em, this is the old code,
could you give some advice about it?

Comment (minggo) [reply cocos/base/ccUTF8.h]: I didn't notice that it is old code. I have not idea too.

File: cocos/base/ccUTF8.cpp [hunk 3]
```diff
@@ -380,11 +381,27 @@ void StringUTF8::replace(const std::string& newStr)
...
+{
     std::string charSequence;
+    size_t maxLen = _str.size() - pos;
+    if (len > maxLen)
+    {
+        len = maxLen;
+    }
 
-    for (auto& charUtf8 : _str)
+    len += pos;
```
Comment (minggo) [inline cocos/base/ccUTF8.cpp]: I think it is better to use another variable to make it more clear, such as 
```c++
auto endPos = pos + len;
```

Comment (summerinsects) [reply cocos/base/ccUTF8.cpp]: maybe `endPos = _str.size();` is better. I modify it

File: cocos/base/ccUTF8.cpp [hunk 4]
```diff
@@ -380,11 +381,27 @@ void StringUTF8::replace(const std::string& newStr)
...
+{
     std::string charSequence;
+    std::size_t maxLen = _str.size() - pos;
+    if (len > maxLen)
+    {
+        len = maxLen;
+    }
 
-    for (auto& charUtf8 : _str)
+    std::size_t endPos = _str.size();
```
Comment (minggo) [inline cocos/base/ccUTF8.cpp]: endPos is related to len

Comment (summerinsects) [reply cocos/base/ccUTF8.cpp]: I am sorry. I fixed it

**Review by minggo — APPROVED** [review]

File: cocos/base/ccUTF8.h (hunk 2 above)
Comment (perminovVS) [reply cocos/base/ccUTF8.h]: StringUTF8 it my api, for ease work cursor in TextField and other.
See specification UTF8 https://tools.ietf.org/html/rfc3629
UTF-8, the object of this memo, has a one-octet encoding unit.  It
   uses all bits of an octet, but has the quality of preserving the full
   US-ASCII [US-ASCII] range: US-ASCII characters are encoded in one
   octet having the normal US-ASCII value, and any octet with such a
   value can only stand for a US-ASCIIUTF-8, the object of this memo, has a one-octet encoding unit.  It
   uses all bits of an octet, but has the quality of preserving the full
   US-ASCII [US-ASCII] range: US-ASCII characters are encoded in one
   octet having the normal US-ASCII value, and any octet with such a
   value can only stand for a US-ASCII character, and nothing else. character, and nothing else.

Right name it function - isASCII.
In Russian ASCII encoding for Window named alias - ansi.

Sorry for the confusion.
Possible name:
isASCII
isOneByteChar
isOneByte

**Pull request merged**